pkg/apply/task/inv_set_task.go

```diff
@@ -152,24 +154,44 @@ func (i *DeleteOrUpdateInvTask) updateInventory(taskContext *taskrunner.TaskCont
 
 	// If an object is invalid and was previously stored in the inventory,
 	// then keep it in the inventory so it can be applied/pruned next time.
-	invalidObjects := i.PrevInventory.Intersection(taskContext.InvalidObjects())
+	invalidObjects := prevInventory.Intersection(taskContext.InvalidObjects())
 	klog.V(4).Infof("keep in inventory %d invalid objects", len(invalidObjects))
 	invObjs = invObjs.Union(invalidObjects)
 
 	klog.V(4).Infof("get the apply status for %d objects", len(invObjs))
-	objStatus := taskContext.InventoryManager().Inventory().Status.Objects
+	objStatus := taskContext.InventoryManager().Inventory().ObjStatuses
 
 	klog.V(4).Infof("set inventory %d total objects", len(invObjs))
-	err := i.InvClient.Replace(taskContext.Context(), i.InvInfo, invObjs, objStatus, i.DryRun)
+	// Skip entire function for dry-run.
+	if i.DryRun.ClientOrServerDryRun() {
+		klog.V(4).Infoln("dry-run replace inventory object: not applied")
+		return nil
+	}
+
+	i.ClusterInventory.SetObjects(invObjs)
+	if err := i.InvClient.CreateOrUpdate(taskContext.Context(), i.ClusterInventory, inventory.UpdateOptions{}); err != nil {
```

Comment:
I think we might want to change this interface to take a method to mutate the object. That way it can handle RetryOnConflict.

Reply:
Hmm, will taking a mutate method work for a sharded client?
I guess it should, right?